src/main/java/com/aws/greengrass/builtin/services/pubsub/SubscriptionTrie.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Object cb) {
+        return remove(topic, Collections.singleton(cb));
+    }
+
+    /**
+     * Remove entry for a set of callbacks.
+     *
+     * @param topic topic
+     * @param cbs   callbacks
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Set<Object> cbs) {
+        SubscriptionTrie sub = lookup(topic);
+        if (sub == null) {
+            return false;
+        }
+        return sub.callbacks.removeAll(cbs);
+    }
+
+    /**
+     * Size of trie.
+     *
+     * @return size
+     */
+    public int size() {
+        int[] size = {this.callbacks.size()};
+        children.forEach((s, t) -> {
+            size[0] += t.size();
+        });
+        return size[0];
+    }
+
+    /**
+     * Add a topic callback.
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return true
+     */
+    public boolean add(String topic, Object cb) {
+        ConcurrentHashMap.KeySetView<Object, Boolean> cbs = ConcurrentHashMap.newKeySet();
+        cbs.add(cb);
+        put(topic, cbs);
+        return true;
+    }
+
+    /**
+     * Add a topic and a set of callbacks.
+     *
+     * @param topic topic
+     * @param cbs   callbacks
+     */
+    public void put(String topic, Set<Object> cbs) {
+        SubscriptionTrie[] current = {this};
+        for (String topicLevel : topic.split(TOPIC_LEVEL_SEPARATOR)) {
+            current[0] = current[0].children.computeIfAbsent(topicLevel, SubscriptionTrie::new);
+        }
+        current[0].callbacks.addAll(cbs);
+    }
+
+    private Set<SubscriptionTrie> getMatchingPaths(String topicLevel, Set<Object> result) {
+        Set<SubscriptionTrie> paths = new LinkedHashSet<>();
+
+        SubscriptionTrie childPath = this.children.get(topicLevel);
+        if (childPath != null) {
+            paths.add(childPath);
+        }
+
+        SubscriptionTrie childPlusPath = this.children.get(SINGLE_LEVEL_WILDCARD);
+        if (childPlusPath != null) {
+            paths.add(childPlusPath);
+        }
+
+        SubscriptionTrie childPoundPath = this.children.get(MULTI_LEVEL_WILDCARD);
+        if (childPoundPath != null) {
+            paths.add(childPoundPath);
+            result.addAll(childPoundPath.callbacks);
+        }
+
+        return paths;
+    }
+
+    /**
+     * Get callback objects given a topic.
+     *
+     * @param topic topic
+     * @return a set of callback objects
+     */
+    public Set<Object> get(String topic) {
+        String[] topicLevels = topic.split(TOPIC_LEVEL_SEPARATOR);
+        Set<Object> result = new LinkedHashSet<>();
+        Set<SubscriptionTrie> paths = this.getMatchingPaths(topicLevels[0], result);
+
+        for (int iter = 1; iter < topicLevels.length && !paths.isEmpty(); iter++) {
+            Set<SubscriptionTrie> newPaths = new LinkedHashSet<>();
```

Comment:
Why create a new set here? Why not just append to the result set?

Reply:
Not sure how this can be done. the new set is for storing all current matching paths and result is for callbacks

Reply:
You already have a result set, right? So why don't you just add to that set instead of creating a new set, adding to that set, then adding that set into the result set

Reply:
So make `getMatchingPaths` take another argument `newPaths`. Instead of returning the new paths from `getMatchingPaths `, just put them in the given newPaths argument. Similar to how the `result` argument work.

Nit: consider rename `iter` -> `level`?